### PR TITLE
Add tests and automated publishing the package to NPM

### PR DESCRIPTION
**Pull request by robmosca**
This PR:
- introduces automated publishing of the package using the changeset action. The setup is identical to the one in single-spa-react
- introduces Jest tests on the API

For automated publishing to work, it is necessary that:

- The `GITHUB_TOKEN` is granted permissions to create and approve pull requests. This can be done in `Settings` -> `Actions` -> `General` -> `Workflow permissions` -> `Allow GitHub Actions to create and approve pull requests`
- The admin of the repo creates an `NPM_TOKEN` secret with an NPM token with write permissions to the organization (or at least the corresponding package, for publishing).

### CODE REVIEW

File: .changeset/config.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "$schema": "https://unpkg.com/@changesets/config@2.3.1/schema.json",
+  "changelog": [
+    "@changesets/changelog-github",
+    { "repo": "single-spa/single-spa-react" }
```
Comment (joeldenning) [inline .changeset/config.json:5]:
```suggestion
    { "repo": "single-spa/import-map-overrides }
```

File: .changeset/plenty-nails-lie.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"import-map-overrides": patch
+---
+
+Add automation to publish package
```
Comment (joeldenning) [inline .changeset/plenty-nails-lie.md:5]: This seems redundant to `automate release flow` in the `fast-chairs-guess.md` changeset. Let's remove this changeset?

File: .github/workflows/build-test-release.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      - name: Checkout Repo
+        uses: actions/checkout@v3
+        with:
+          # check out full history
+          fetch-depth: 0
+
+      - name: Install Pnpm
+        uses: pnpm/action-setup@v2.0.1
+        with:
+          version: 8.6.9
```
Comment (joeldenning) [inline .github/workflows/build-test-release.yml:24]:
```suggestion
          version: 8.8.0
```

File: .gitignore [hunk 4]
```diff
@@ -65,4 +65,4 @@ dist
 /.idea
 
 # VSCode IDE files
-.vscode
+/.vscode
```
Comment (joeldenning) [inline .gitignore:68]: What's the reason for this change? Do we want nested `.vscode` files in this repo?

File: rollup.config.js [hunk 5]
```diff
@@ -70,6 +70,7 @@ export default [
     },
     plugins: [
       babel({
+        babelHelpers: "bundled",
```
Comment (joeldenning) [inline rollup.config.js:73]: Why is this needed? Are we using generator syntax or something else that requires babel runtime?

File: rollup.config.js [hunk 6]
```diff
@@ -80,6 +81,7 @@ export default [
           ],
         ],
       }),
+      nodeResolve(),
```
Comment (joeldenning) [inline rollup.config.js:84]: Why do we need this plugin now, but didn't before?

File: src/api/js-api.js [hunk 7]
```diff
@@ -207,7 +207,13 @@ function init() {
               if (scriptEl.src) {
                 nextPromise = fetchExternalMap(scriptEl.src);
               } else {
-                nextPromise = Promise.resolve(JSON.parse(scriptEl.textContent));
+                let parsedContent;
+                try {
+                  parsedContent = JSON.parse(scriptEl.textContent);
+                } catch (e) {
```
Comment (joeldenning) [inline src/api/js-api.js:213]: Is swallowing this json parse error ok? The user should be aware if their import map is invalid.

File: src/api/js-api.js [hunk 8]
```diff
@@ -341,8 +347,8 @@ function init() {
         externalOverrideMapPromises[importMapUrl] ||
         (externalOverrideMapPromises[importMapUrl] =
           fetchExternalMap(importMapUrl));
-      return promise.then(() =>
-        includes(imo.invalidExternalMaps, importMapUrl)
+      return promise.then(
+        () => !includes(imo.invalidExternalMaps, importMapUrl)
```
Comment (joeldenning) [inline src/api/js-api.js:351]: Is this a bug that you fixed? Please link issue if there was one for this

File: src/api/js-api.js [hunk 9]
```diff
@@ -367,6 +373,7 @@ function init() {
 
   function fireEvent(type) {
     // Set timeout so that event fires after the change has totally finished
+
```
Comment (joeldenning) [inline src/api/js-api.js:376]:
```suggestion
```

File: src/api/localStorageMock.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+// Mocks for the localStorage
```
Comment (joeldenning) [inline src/api/localStorageMock.js:1]: Seems redundant to have `local-storage-mock.js` and `localStorageMock.js`

File: src/server/server-api.js [hunk 11]
```diff
@@ -1,4 +1,4 @@
-import cookie from "cookie";
+const cookie = require("cookie");
```
Comment (joeldenning) [inline src/server/server-api.js:1]: Why the change to CJS?

File: src/util/includes.test.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
+import { includes } from "./includes";
```
Comment (joeldenning) [inline src/util/includes.test.js:1]: I think we have our own implementation of `includes()` to support IE11 - we should switch to Array.prototype.includes at some point, but maybe not now to avoid a breaking change.

**Review by joeldenning — COMMENTED** [review]
Thanks for your work on this, these tests will help a lot. I've left a few comments and questions

Do I need to add an NPM_TOKEN environment variable to this repo before merging it?

File: .changeset/config.json (hunk 1 above)
Comment (robmosca) [reply .changeset/config.json]: Sorry about this. Fixed.

File: .changeset/plenty-nails-lie.md (hunk 2 above)
Comment (robmosca) [reply .changeset/plenty-nails-lie.md]: Of course. This PR has been reworked so many times that with iterations (it has been reopened quite a few times) some comments went lost and some inconsistencies crept in. Removing this file.

File: .gitignore (hunk 4 above)
Comment (robmosca) [reply .gitignore]: We don't want them, but I don't think VS Code would actually create nested `.vscode` folders?
I made it to harmonize with the line above, for JetBrain IDE files.
In any case, reverting the change. 👍🏽

**Review by robmosca — COMMENTED** [review]
Addressed the comments.

**Comment by robmosca** [comment]
> Thanks for your work on this, these tests will help a lot. I've left a few comments and questions
> 
> Do I need to add an NPM_TOKEN environment variable to this repo before merging it?

For automated publishing to work, you will need to do the following, before merging, @joeldenning:

- The `GITHUB_TOKEN` is granted permissions to create and approve pull requests. This can be done in `Settings` -> `Actions` -> `General` -> `Workflow permissions` -> `Allow GitHub Actions to create and approve pull requests`
- The admin of the repo creates an `NPM_TOKEN` secret with an NPM token with write permissions to the organization (or at least the corresponding package, for publishing).

File: src/server/server-api.js (hunk 11 above)
Comment (joeldenning) [reply src/server/server-api.js:1]: I am hesitant about this change because it could impact a lot of things (bundling, tests, linter, etc). All ESM is better.

I read through the rollup documentation linked to in the warning you shared (https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency) and the solution seems to be to add `@rollup/plugin-node-resolve` rather than to change it to a CJS import. Is there an error even when rollup node resolve is present?

Have you tested that import-map-overrides works server-side after this change?

**Review by joeldenning — COMMENTED** [review]
Looks good, one last comment

**Comment by robmosca** [comment]
> I read through the rollup documentation linked to in the warning you shared (https://rollupjs.org/guide/en/#warning-treating-module-as-external-dependency) and the solution seems to be to add @rollup/plugin-node-resolve rather than to change it to a CJS import. Is there an error even when rollup node resolve is present?
> 
> Have you tested that import-map-overrides works server-side after this change?

To unblock this PR I have reverted all changes related to the warning I get when building `import-map-overrides-server.js`.
I tracked the problem in [this new issue](https://github.com/single-spa/import-map-overrides/issues/99) and I can tackle it in a separate PR.

**Review by joeldenning — APPROVED** [review]
